# **References**

* [15 basic step for NLP](https://github.com/themadan/12.NLP-ear-and-tongue-sound-and-text-/blob/master/15_natural_language_processing.ipynb)

* [Microsoft developer](https://devblogs.microsoft.com/cse/2015/11/29/emotion-detection-and-recognition-from-text-using-deep-learning/)
* [Medium](https://medium.com/the-research-nest/applied-machine-learning-part-3-3fd405842a18) <br> [Work](https://github.com/aditya-xq/Text-Emotion-Detection-Using-NLP)
* [Compete web application](https://github.com/maelfabien/Multimodal-Emotion-Recognition)

* [git](https://github.com/Harsh24893/EmotionRecognition)
* [Notebook](https://github.com/abishekarun/Text-Emotion-Classification/blob/master/emotion_classification.ipynb)



# **Import Packages**

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns

from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from scipy.stats import itemfreq
from sklearn.feature_extraction.text import CountVectorizer,TfidfTransformer,HashingVectorizer
from sklearn.pipeline import Pipeline
import nltk
from nltk.corpus import stopwords
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import confusion_matrix




/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [3]:
from google.colab import drive
drive.mount('/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /drive


In [0]:
columns=['emotion','content']
data = pd.read_csv('/drive/My Drive/Fusemachines Nepal/NLP/ISEAR.csv',names=columns)

In [6]:
data.head()

emotion                                            content
0      joy  On days when I feel close to my partner and ot...
1     fear  Every time I imagine that someone I love or I ...
2    anger  When I had been obviously unjustly treated and...
3  sadness  When I think about the short time that we live...
4  disgust  At a gathering I found myself involuntarily si...

In [7]:
data.describe()

emotion                    content
count     7446                       7446
unique       7                       7379
top        joy  When my grandfather died.
freq      1082                          8

* We have 7 emotion category.
* We have Total of 7446 data.
* Class joy has the highest number of data.

In [8]:
data.shape

(7446, 2)

In [9]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7446 entries, 0 to 7445
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   emotion  7446 non-null   object
 1   content  7446 non-null   object
dtypes: object(2)
memory usage: 174.5+ KB


## **Number of data for each emotion.**

In [10]:
data.emotion.value_counts()

joy        1082
sadness    1074
anger      1069
fear       1063
shame      1059
disgust    1059
guilt      1040
Name: emotion, dtype: int64

* We have 7 emotion types
* The data seems to be quite balanced

# **Clean Text**

In [11]:
data['content'][0]
print(data['content'][0])
print(data['content'].str.len())

On days when I feel close to my partner and other friends.   
When I feel at peace with myself and also experience a close  
contact with people whom I regard greatly.
0       167
1        92
2        88
3       139
4       144
       ... 
7441    181
7442     72
7443    875
7444    358
7445     90
Name: content, Length: 7446, dtype: int64


## **Remove all the new line characters**

In [12]:
data['content'] = data['content'].str.replace('\n', '')
print(data['content'][0])
print(data['content'].str.len())

On days when I feel close to my partner and other friends.   When I feel at peace with myself and also experience a close  contact with people whom I regard greatly.
0       165
1        91
2        87
3       137
4       142
       ... 
7441    179
7442     71
7443    862
7444    353
7445     89
Name: content, Length: 7446, dtype: int64


## **Replace full stop with blank**

In [13]:
data['content'] = data['content'].str.replace('.', '')
print(data['content'][0])
print(data['content'].str.len())

On days when I feel close to my partner and other friends   When I feel at peace with myself and also experience a close  contact with people whom I regard greatly
0       163
1        90
2        86
3       136
4       141
       ... 
7441    177
7442     70
7443    853
7444    349
7445     88
Name: content, Length: 7446, dtype: int64


## **Remove irrelevant characters other than alphanumeric and space**

In [14]:
data['content']=data['content'].str.replace('[^A-Za-z0-9\s]+', '')
print(data['content'][0])
print(data['content'].str.len())

On days when I feel close to my partner and other friends   When I feel at peace with myself and also experience a close  contact with people whom I regard greatly
0       163
1        89
2        86
3       136
4       141
       ... 
7441    177
7442     70
7443    852
7444    348
7445     88
Name: content, Length: 7446, dtype: int64


## **Remove links from the text**

In [15]:
data['content']=data['content'].str.replace('http\S+|www.\S+', '', case=False)
print(data['content'][0])
print(data['content'].str.len())

On days when I feel close to my partner and other friends   When I feel at peace with myself and also experience a close  contact with people whom I regard greatly
0       163
1        89
2        86
3       136
4       141
       ... 
7441    177
7442     70
7443    852
7444    348
7445     88
Name: content, Length: 7446, dtype: int64


## **Convert everything to lowercase**

In [16]:
data['content']=data['content'].str.lower()
print(data['content'][0])
print(data['content'].str.len())

on days when i feel close to my partner and other friends   when i feel at peace with myself and also experience a close  contact with people whom i regard greatly
0       163
1        89
2        86
3       136
4       141
       ... 
7441    177
7442     70
7443    852
7444    348
7445     88
Name: content, Length: 7446, dtype: int64


## **Removing Punctuation, Symbols**

In [17]:
data['content'] = data['content'].str.replace('[^\w\s]',' ')
print(data['content'][0])
print(data['content'].str.len())

on days when i feel close to my partner and other friends   when i feel at peace with myself and also experience a close  contact with people whom i regard greatly
0       163
1        89
2        86
3       136
4       141
       ... 
7441    177
7442     70
7443    852
7444    348
7445     88
Name: content, Length: 7446, dtype: int64


## **Removing Stop Words using NLTK**

In [19]:
from nltk.corpus import stopwords

nltk.download('stopwords')
stop = stopwords.words('english')
data['content'] = data['content'].apply(lambda x: " ".join(x for x in x.split() if x not in stop))

print(data['content'][0])
print(data['content'].str.len())

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
days feel close partner friends feel peace also experience close contact people regard greatly
0        94
1        72
2        50
3        62
4       102
       ... 
7441    110
7442     29
7443    474
7444    226
7445     48
Name: content, Length: 7446, dtype: int64


## **Lemmatisation**

In [21]:
from textblob import Word

nltk.download('wordnet')
data['content'] = data['content'].apply(lambda x: " ".join([Word(word).lemmatize() for word in x.split()]))
print(data['content'][0])
print(data['content'].str.len())

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.
day feel close partner friend feel peace also experience close contact people regard greatly
0        92
1        72
2        50
3        61
4       101
       ... 
7441    109
7442     29
7443    468
7444    225
7445     47
Name: content, Length: 7446, dtype: int64


## **Correcting Letter Repetitions**

In [22]:
import re
def de_repeat(text):
    pattern = re.compile(r"(.)\1{2,}")
    return pattern.sub(r"\1\1", text)

data['content'] = data['content'].apply(lambda x: " ".join(de_repeat(x) for x in x.split()))

print(data['content'][0])
print(data['content'].str.len())

day feel close partner friend feel peace also experience close contact people regard greatly
0        92
1        72
2        50
3        61
4       101
       ... 
7441    109
7442     29
7443    468
7444    225
7445     47
Name: content, Length: 7446, dtype: int64


# **Assign Target Variable**

In [23]:
target=data.emotion
data = data.drop(['emotion'],axis=1)
print(target)
print(data)

0           joy
1          fear
2         anger
3       sadness
4       disgust
         ...   
7441      anger
7442    sadness
7443    disgust
7444      shame
7445      guilt
Name: emotion, Length: 7446, dtype: object
                                                content
0     day feel close partner friend feel peace also ...
1     every time imagine someone love could contact ...
2     obviously unjustly treated possibility elucida...
3     think short time live relate period life think...
4     gathering found involuntarily sitting next two...
...                                                 ...
7441  last week planned play tennis booked tennis co...
7442                      ill stay hospital period time
7443  day back waiting bus bus stop getting bus prep...
7444  day back tutorial class teacher randomly assig...
7445    quarrelled sister deliberately messed belonging

[7446 rows x 1 columns]


# **LabelEncoder for target**

In [24]:
le=LabelEncoder()
target=le.fit_transform(target)
print(target)

[4 2 0 ... 1 6 3]


# **Split Data into train & test**

In [25]:
X_train, X_test, y_train, y_test = train_test_split(data,target,stratify=target,test_size=0.4, random_state=42)
print(X_train)
print(y_train)
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)



                                                content
504                went home train sicilia molested man
2282  met girl 22 year old married liked asked date ...
3476          sad first boyfriend finished relationship
6478                 heard friend started drinking beer
2250  possibility getting better professional life v...
...                                                 ...
4074  aunt phoned ask refused invitation dinner home...
6622  wet head bed one day sister discovered reporte...
2758  feeling unable preserve one idea ambition inno...
1465  angry several driver showed aggressive dangero...
2824  possibility act certain activity better done r...

[4467 rows x 1 columns]
[1 1 5 ... 5 0 3]
(4467, 1)
(4467,)
(2979, 1)
(2979,)


# **Tokenization**

Tokenization can be done in a variety of ways, namely 
* Bag of words, 
* tf-idf, Glove, 
* word2vec ,
* fasttext etc. Lets see how they can be applied and how they affect the accuracy

## **Bag of Words and Naive Bayes**

In [26]:
count_vect = CountVectorizer()
X_train_counts = count_vect.fit_transform(X_train.content)
X_test_counts =count_vect.transform(X_test.content)
print('Shape of Term Frequency Matrix: ',X_train_counts.shape)

Shape of Term Frequency Matrix:  (4467, 6372)


### Naive Bayes Model

In [27]:
clf = MultinomialNB().fit(X_train_counts,y_train)
predicted = clf.predict(X_test_counts)
nb_clf_accuracy = np.mean(predicted == y_test) * 100
print(nb_clf_accuracy)

55.58912386706949


## **Pipeline and Naive Bayes**

In [0]:
def print_acc(model):
    predicted = model.predict(X_test.content)
    accuracy = np.mean(predicted == y_test) * 100
    print(accuracy)

In [29]:
nb_clf = Pipeline([('vect', CountVectorizer()), ('clf', MultinomialNB())])
nb_clf = nb_clf.fit(X_train.content,y_train)
print_acc(nb_clf)

55.58912386706949


## **TF IDF transformer and Naive Bayes**

In [30]:
nb_clf = Pipeline([('vect', CountVectorizer()), ('tfidf', TfidfTransformer()), ('clf', MultinomialNB())])
nb_clf = nb_clf.fit(X_train.content,y_train)
print_acc(nb_clf)

55.052030882846594
